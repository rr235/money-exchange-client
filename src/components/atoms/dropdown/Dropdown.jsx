import React, { useState } from 'react';
import classNames from 'classnames';
import styles from './dropdown.styles.scss';

const getSelection = (options, selectedValue) => {
  const selection = options.filter((option) => option.value === selectedValue);
  return selection.length ? selection[0].text : options[0].text;
};

const DropdownListItem = ({ value, text, selected, onClick }) => {
  const onClickHandler = () => {
    onClick({ value, text });
  };

  return (
    <li
      data-value={value}
      role="option"
      aria-selected={selected}
      key={value}
      onClick={onClickHandler}
      className={styles.option}
    >
      {text}
    </li>
  );
};

const Dropdown = ({ options, onSelect, className, selectedValue }) => {
  const [showOptions, setShowOptions] = useState(false);

  const optionsClassName = classNames(styles.options, {
    [styles.show]: showOptions,
  });
  const dropdownClassName = classNames(styles.dropdown, className);

  const onOptionSelectHandler = ({ text, value }) => {
    setShowOptions(false);
    onSelect({ text, value });
  };

  const onClickHandler = () => {
    // toggle options
    setShowOptions(!showOptions);
  };

  return (
    <div className={dropdownClassName}>
      <button
        type="button"
        aria-haspopup="listbox"
        onClick={onClickHandler}
        className={styles.button}
      >
        {getSelection(options, selectedValue)}
      </button>
      <ul role="listbox" className={optionsClassName}>
        {options.map(({ value, text, selected }) => (
          <DropdownListItem
            value={value}
            text={text}
            selected={selected}
            key={value}
            onClick={onOptionSelectHandler}
          />
        ))}
      </ul>
    </div>
  );
};

export default Dropdown;
