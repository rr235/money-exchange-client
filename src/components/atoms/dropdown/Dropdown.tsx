import React, { useState } from 'react';
import classNames from 'classnames';
import styles from './dropdown.styles.scss';

type Option = {
  value: number | string;
  text: string;
  selected?: boolean;
  onClick?(value: Option): void;
};

type props = {
  className?: string;
  options: Option[];
  onSelect?(value: Option): void;
};

const getSelection = (options: Option[]): string => {
  const selection = options.filter((option) => option.selected);
  return selection.length ? selection[0].text : options[0].text;
};

const DropdownListItem = ({ value, text, selected, onClick }: Option) => {
  const onClickHandler = () => {
    onClick({ value, text });
  };

  return (
    <li
      data-value={value}
      role="option"
      aria-selected={selected}
      key={value}
      onClick={onClickHandler}
    >
      {text}
    </li>
  );
};

const Dropdown = ({ options, onSelect, className }: props) => {
  const [showOptions, setShowOptions] = useState<boolean>(false);

  const optionsClassName = classNames(styles.options, {
    [styles.show]: showOptions,
  });
  const buttonClassName = classNames(styles.button, className);

  const onOptionSelectHandler = ({ text, value }: Option) => {
    onSelect({ text, value });
  };

  const onClickHandler = () => {
    // toggle options
    setShowOptions(!showOptions);
  };

  return (
    <>
      <button
        type="button"
        aria-haspopup="listbox"
        onClick={onClickHandler}
        className={buttonClassName}
      >
        {getSelection(options)}
      </button>
      <ul role="listbox" className={optionsClassName}>
        {options.map(({ value, text, selected }: Option) => (
          <DropdownListItem
            value={value}
            text={text}
            selected={selected}
            key={value}
            onClick={onOptionSelectHandler}
          />
        ))}
      </ul>
    </>
  );
};

export default Dropdown;
