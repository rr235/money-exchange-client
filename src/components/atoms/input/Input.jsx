import React, { useState, FormEvent } from 'react';
import classNames from 'classnames';
import styles from './input.styles.scss';

const Input = ({ label, id, onChange, className }) => {
  const [input, setInput] = useState('');

  const onChangeHandler = (e) => {
    setInput(e.currentTarget.value);
    if (onChange) {
      onChange(e);
    }
  };

  return (
    <>
      <label htmlFor={id} className={classNames(styles.label, className)}>
        {label}
      </label>
      <input
        type="number"
        id={id}
        className={styles.input}
        value={input}
        onChange={onChangeHandler}
      />
    </>
  );
};

export default Input;
