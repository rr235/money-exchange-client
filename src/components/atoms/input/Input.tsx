import React, { useState, FormEvent } from 'react';
import classNames from 'classnames';
import styles from './input.styles.scss';

type Props = {
  id: string;
  label: string;
  className?: string;
  onChange?(e: FormEvent<HTMLInputElement>): () => void;
};

const Input = ({ label, id, onChange, className }: Props) => {
  const [input, setInput] = useState('');

  const onChangeHandler = (e: FormEvent<HTMLInputElement>): void => {
    setInput(e.currentTarget.value);
    if (onChange) {
      onChange(e);
    }
  };

  return (
    <>
      <label htmlFor={id} className={classNames(styles.label, className)}>
        {label}
      </label>
      <input
        type="number"
        id={id}
        className={styles.input}
        value={input}
        onChange={onChangeHandler}
      />
    </>
  );
};

export default Input;
