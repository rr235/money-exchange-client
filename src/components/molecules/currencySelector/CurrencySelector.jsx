import React from 'react';
import { func, string, shape, number, arrayOf, oneOfType } from 'prop-types';
import Input from '../../atoms/input';
import Dropdown, { optionsShape } from '../../atoms/dropdown';
import styles from './currencySelector.styles.scss';

const CurrencySelector = ({
  options,
  onSelect,
  onChange,
  label,
  id,
  selectedValue,
  inputValue,
}) => (
  <div>
    {options.length && (
      <Dropdown
        options={options}
        onSelect={onSelect}
        id={`select-${id}`}
        className={styles.dropdown}
        selectedValue={selectedValue.code}
      />
    )}
    <span className={styles.balance}>
      {'Balance: '}
      {selectedValue.symbol}
      {selectedValue.balance}
    </span>
    <Input
      label={label}
      id={`input-${id}`}
      onChange={onChange}
      type="number"
      value={inputValue}
    />
  </div>
);

CurrencySelector.propTypes = {
  options: arrayOf(optionsShape).isRequired,
  id: string.isRequired,
  label: string,
  onSelect: func,
  onChange: func,
  selectedValue: shape({
    name: string,
    code: string,
    symbol: string,
    balance: number,
  }),
  inputValue: oneOfType([string, number]),
};

CurrencySelector.defaultProps = {
  onSelect: () => {},
  onChange: () => {},
  label: '',
  selectedValue: {},
  inputValue: null,
};

export default CurrencySelector;
