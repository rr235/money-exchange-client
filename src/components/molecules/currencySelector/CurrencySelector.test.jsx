import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import CurrencySelector from './CurrencySelector';
import Dropdown from '../../atoms/dropdown';
import Input from '../../atoms/input';

const options = [
  { text: 'option1', value: 'val1' },
  { text: 'option2', value: 'val2' },
  { text: 'option3', value: 'val3' },
];

const selectedValue = {
  symbol: 'E',
  code: 'val2',
  balance: 10,
};

describe('CurrencySelector', () => {
  let component;
  beforeAll(() => {
    act(() => {
      component = mount(
        <CurrencySelector
          options={options}
          id="id"
          label="label"
          selectedValue={selectedValue}
          inputValue={10}
        />
      );
    });
  });

  it('should render correctly', () => {
    const tree = renderer
      .create(
        <CurrencySelector
          options={options}
          id="id"
          label="label"
          selectedValue={selectedValue}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should show one dropdown', () => {
    expect(component.find(Dropdown)).toHaveLength(1);
  });

  it('should show one input field', () => {
    expect(component.find(Input)).toHaveLength(1);
  });

  it('should show correct selection', () => {
    const dropdownElement = component.find(Dropdown).find('button');
    expect(dropdownElement.text()).toBe('option2');
  });

  it('should show correct input value', () => {
    const inputElement = component.find(Input).find('input');
    expect(inputElement.prop('value')).toBe(10);
  });

  it('should show exceeds balance message when exceeds balance is true', () => {
    let exceedsBalanceMessage = component.find('.exceedsBalanceMessage').text();
    expect(exceedsBalanceMessage).toBe('');

    act(() => {
      component.setProps({ exceedsBalance: true });
    });
    component.update();

    exceedsBalanceMessage = component.find('.exceedsBalanceMessage').text();
    expect(exceedsBalanceMessage).toBe('Exceeds Balance');
  });
});
