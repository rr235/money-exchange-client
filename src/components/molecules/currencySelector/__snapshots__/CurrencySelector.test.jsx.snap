// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencySelector should render correctly (type="number") 1`] = `
<div>
  <div
    className="dropdown dropdown"
  >
    <button
      aria-haspopup="listbox"
      className="button"
      id="select-from"
      onClick={[Function]}
      type="button"
    >
      option1
    </button>
    <ul
      className="options"
      role="listbox"
    >
      <li
        aria-selected={false}
        className="option"
        data-value="val1"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="option"
        tabIndex="0"
      >
        option1
      </li>
      <li
        aria-selected={false}
        className="option"
        data-value="val2"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="option"
        tabIndex="0"
      >
        option2
      </li>
      <li
        aria-selected={false}
        className="option"
        data-value="val3"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="option"
        tabIndex="0"
      >
        option3
      </li>
    </ul>
  </div>
  <span
    className="balance"
  >
    Balance: 
  </span>
  <label
    className="label"
    htmlFor="input-from"
  >
    From
  </label>
  <input
    autoComplete="off"
    className="input"
    id="input-from"
    onChange={[Function]}
    type="text"
    value={null}
  />
  <div
    className="exceedsBalanceMessage"
  />
</div>
`;
